controllers/volumereplicationgroup_controller.go

```diff
@@ -199,38 +203,159 @@ func (v *VolumeReplicationGroupReconciler) Reconcile(ctx context.Context, req ct
 
 	log.Info("Processing VolumeReplicationGroup ", volRepGroup.Spec, " in ns: ", req.NamespacedName)
 
-	pvcList, err := v.HandlePersistentVolumeClaims(ctx, volRepGroup)
+	// examine the DeletionTimestamp to determine whether the object is under deletion
+	if !volRepGroup.GetDeletionTimestamp().IsZero() {
+		if err := v.runAndRemoveFinalizer(ctx, volRepGroup, vrgFinalizerName); err != nil {
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		// stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
+	// Add finalizer if not present
+	if err := v.addFinalizer(ctx, volRepGroup, vrgFinalizerName); err != nil {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	pvcList, err := v.listPersistentVolumeClaims(ctx, volRepGroup)
 	if err != nil {
 		log.Error(err, "Handling of Persistent Volume Claims of application failed")
 
 		return ctrl.Result{}, err
 	}
 
 	requeue := false
+	requeue = v.handlePersistentVolumeClaims(ctx, volRepGroup, pvcList)
 
-	for _, pvc := range pvcList.Items {
-		volRep := &volrep.VolumeReplication{}
+	if requeue {
+		return ctrl.Result{Requeue: requeue}, nil
+	}
 
-		err = v.Get(ctx, types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}, volRep)
-		if err != nil {
-			if errors.IsNotFound(err) {
-				if err = v.createVolumeReplicationCRForPVC(ctx, pvc); err == nil {
-					continue
-				}
-			}
+	return ctrl.Result{}, nil
+}
 
-			// requeue on failure to ensure any PVC not having a corresponding VR CR
-			requeue = true
+// It might be better move the helper functions like these to a separate
+// package or a separate go file?
+func containsString(values []string, s string) bool {
+	for _, item := range values {
+		if item == s {
+			return true
+		}
+	}
+
+	return false
+}
+
+func removeString(values []string, s string) (result []string) {
```

Comment:
(nit) Name result parameters are confusing and there is no reason to not declare a variable inside the function.  This looks weird.

Reply:
ACK. Will change it.